will/backends/io_adapters/slack.py

```diff
@@ -142,99 +188,123 @@ def normalize_incoming_event(self, event):
                 original_incoming_event=clean_for_pickling(event),
             )
             return m
-        else:
-            # An event type the slack ba has no idea how to handle.
-            pass
+        # An event type the slack ba has no idea how to handle.
+        return None
 
     def set_topic(self, event):
-        headers = {'Accept': 'text/plain'}
+        """Sets the channel topic.  This doesn't actually work anymore since Slack has removed that ability from bots.
+        Leaving the code here in case they re-enable it."""
         data = self.set_data_channel_and_thread(event)
-        data.update({
-            "token": settings.SLACK_API_TOKEN,
-            "as_user": True,
-            "topic": event.content,
-        })
-        if data["channel"].startswith("G"):
-            url = SLACK_PRIVATE_SET_TOPIC_URL
-        else:
-            url = SLACK_SET_TOPIC_URL
-        r = requests.post(
-            url,
-            headers=headers,
-            data=data,
-            **settings.REQUESTS_OPTIONS
+        self.client._web_client.api_call(  # pylint: disable=protected-access
+            "conversations.setTopic",
+            topic=event.content,
+            channel=data["channel"],
         )
-        self.handle_request(r, data)
-
-    def handle_outgoing_event(self, event):
-        if event.type in ["say", "reply"]:
-            if "kwargs" in event and "html" in event.kwargs and event.kwargs["html"]:
-                event.content = SlackMarkdownConverter().convert(event.content)
-
-            event.content = event.content.replace("&", "&amp;")
-            event.content = event.content.replace(r"\_", "_")
 
-            kwargs = {}
-            if "kwargs" in event:
-                kwargs.update(**event.kwargs)
-
-            if hasattr(event, "source_message") and event.source_message and "channel" not in kwargs:
-                self.send_message(event)
-            else:
-                # Came from webhook/etc
-                # TODO: finish this.
-                target_channel = kwargs.get("room", kwargs.get("channel", None))
-                if target_channel:
-                    event.channel = self.get_channel_from_name(target_channel)
-                    if event.channel:
+    def send_file(self, event):
```

Comment:
Just getting to a closer look at the slack backend. Let's also remove `send_file` here. We definitely need file upload support so if you're able to submit this as a separate PR we can review separetely and get it merged in before we cut a 2.2.0 release.

Reply:
I think you need to make up your mind here.  You wanted me to pull send_file() out because there are (possibly) several alternative implementations and an existing PR.  That's fine, but *I* won't be submitting a separate PR for this.  Splitting things into multiple PRs just for the sake of multiple PRs adds pointless burden to me.  I'll pull this out, and then you can look at code submitted by someone else.

Reply:
@pastorhudson @rsalmond what are your thoughts here? I'm of the mind that we need to keep this PR limited to rewriting the backend to support current methods and the file upload is a great bonus _however_ I'm wary of bringing in changes that will break other implementations.

Reply:
I get what you’re saying, but I think if we want file support and this works then merge it in. At this point this project is almost dead because of stagnation and lack of merging pull requests. For the health of the project I think we’d be better to merge some stuff even if it’s not quite perfect just to keep the ball rolling. We can always fix bugs if we discover them as long as things are getting regularly merged.

Reply:
Removed